Add Husqvarna Automower integration

Thomas55555 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add Husqvarna Automower integration. This was previously a custom integration. As there is the `lawn_mower` entity now, I thought it's a good moment to integrate it in Home Assistant.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28817

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mikey0000 commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,101 @@
+    )
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})
```
```suggestion
```
No need for this line

mikey0000 commented on tests/components/husqvarna_automower/const.py:
```diff
@@ -0,0 +1,218 @@
+        },
+        "metadata": {"connected": True, "statusTimestamp": 1685899526195},
+        "calendar": {
+            "tasks": [
```
should this be here? looks like config for your own mower?

mikey0000 commented on tests/components/husqvarna_automower/const.py:
```diff
@@ -0,0 +1,218 @@
+
+AUTOMOWER_CONFIG_DATA = {
+    "auth_implementation": "husqvarna_automower",
+    "token": {
```
looks like this should be removed

mikey0000 commented on homeassistant/components/husqvarna_automower/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "config_flow": true,
+  "dependencies": ["application_credentials"],
+  "documentation": "https://www.home-assistant.io/integrations/husqvarna_automower",
+  "homekit": {},
```
```suggestion
```

mikey0000 commented on homeassistant/components/husqvarna_automower/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "homekit": {},
+  "iot_class": "cloud_push",
+  "requirements": ["aioautomower==2023.8.1"],
+  "ssdp": [],
```
```suggestion
```

mikey0000 commented on homeassistant/components/husqvarna_automower/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "iot_class": "cloud_push",
+  "requirements": ["aioautomower==2023.8.1"],
+  "ssdp": [],
+  "zeroconf": []
```
```suggestion
```

mikey0000 commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,125 @@
+
+    _attr_name: str | None = None
+    _attr_supported_features = SUPPORT_STATE_SERVICES
+    _attr_translation_key = "mower"
```
```suggestion
   _attr_translation_key = "mower"
   _attr_has_entity_name = True
```
and add the entity to strings

I believe this is the correct way to do this.

e.g 
```
"entity": {
    "lawn_mower": {
    "mower": {
        "name": "Husqvarna Mower"
      },
    }
    ```

Thomas55555 replied:
And `_attr_has_entity_name` is already in `entity.py`

Thomas55555 replied:
I've removed the `translation_key`. Because the mower (as main entity) should just have the name of the device. And there is no translation required.
`_attr_has_entity_name = True` is already in `entity.py`

mikey0000 commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,101 @@
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```

mikey0000 commented:
I've left a few comments, hopefully helps push this along faster. Though I helped get the lawn mower entity into HA, I am by no means an expert. So take my comments as such. Though most are similar issues that have been raised on my own integrations. Cheers and thanks!

joostlek commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,100 @@
+            name=DOMAIN,
+        )
+        api_key = None
+        ap_storage: dict = cast(dict, hass.data.get("application_credentials"))
+        ap_storage_data: dict = ap_storage[DATA_STORAGE].__dict__["data"]
+        for k in ap_storage_data:
+            api_key = ap_storage_data[k]["client_id"]
+        entry_dict = entry.as_dict()
+        access_token = entry_dict["data"]["token"]
+        scope = entry_dict["data"]["token"]["scope"]
```
I am not sure how at this point, but want to mention it, I think this is data that should be in the entry itself, I'll have a second look later.

Thomas55555 replied:
I had this in the beginning:
```
    implementation = await async_get_config_entry_implementation(hass, entry)
    coordinator = AutomowerDataUpdateCoordinator(
        hass,
        implementation,
        entry=entry,
    )
```
and used in the coordinator `api_key = implementation.client_id`. But the I got problems with the tests I couldn't solved.

joostlek replied:
We will take a look at this when we start with the tests

joostlek commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,100 @@
+        """Fetch data from Husqvarna."""
+        try:
+            await self.session.connect()
+        except AsyncioTimeoutError as error:
+            _LOGGER.debug("Asyncio timeout: %s", error)
+            raise ConfigEntryNotReady from error
```
TimeoutErrors from asyncio are already catched in the data update coordiantor

joostlek commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,100 @@
+            raise ConfigEntryNotReady from error
+        except Exception as error:
+            _LOGGER.debug("Exception in async_setup_entry: %s", error)
+            raise ConfigEntryAuthFailed from Exception
```
Why would a general Exception cause a reauth? Wouldn't having no internet trigger this (as that raises a CannotConnect error or something along those lines) That should raise `UpdateFailed` instead.

Thomas55555 replied:
Yes, that's to general

joostlek commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,100 @@
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle unload of an entry."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    with contextlib.suppress(Exception):
```
What could raise?

Thomas55555 replied:
If there is for example no internet there could be an error.
Or if the session is broken out of any reason, there would be no possibility to unload the integration.
So there could be nearly anything rising.

MartinHjelmare replied:
The library should handle transport errors and  raise library specific errors instead. When using aiohttp as a transport only the aiohttp exceptions will be relevant to catch in the transport layer.

Thomas55555 replied:
> What could raise?

Basically everything. So this just ensures, that the entry is unloaded, whatever happens.

joostlek commented on homeassistant/components/husqvarna_automower/config_flow.py:
```diff
@@ -0,0 +1,57 @@
+
+class HusqvarnaConfigFlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler,
+    data_entry_flow.FlowHandler,
```
```suggestion
```
Shouldn't need this

joostlek commented on homeassistant/components/husqvarna_automower/config_flow.py:
```diff
@@ -0,0 +1,57 @@
+            data=data,
+        )
+
+    async def async_step_reauth(self, user_input: Mapping[str, Any]) -> FlowResult:
+        """Perform reauth upon an API authentication error."""
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(self, user_input=None) -> FlowResult:
+        """Dialog that informs the user that reauth is required."""
+        if user_input is None:
+            _LOGGER.debug("USER INPUT NONE")
+            return self.async_show_form(
+                step_id="reauth_confirm",
+                data_schema=vol.Schema({}),
+            )
+        return await self.async_step_user()
```
If possible, let's move reauth to a follow up PR so we can focus on the rest and make it as small as possible. The reauth still might need some work when looking at logging in with the right account (for example, you create the config with account A, and reauth with acount B). But let's keep that out of scope for now, and instead of ConfigEntryAuthFailed, just raise ConfigEntryNotReady for the time being

joostlek commented on homeassistant/components/husqvarna_automower/config_flow.py:
```diff
@@ -0,0 +1,57 @@
+    async def async_step_finish(self, unique_id, data) -> FlowResult:
+        """Complete the config entries."""
+        await self.async_set_unique_id(unique_id)
+
```
```suggestion
            self._abort_if_unique_id_configured()
```
This will raise an issue if the unique_id is already set. This way you can't login twice with the same account.

joostlek commented on homeassistant/components/husqvarna_automower/config_flow.py:
```diff
@@ -0,0 +1,57 @@
+    DOMAIN = DOMAIN
+    VERSION = 1
+
+    async def async_oauth_create_entry(self, data: dict) -> FlowResult:
```
```suggestion
    async def async_oauth_create_entry(self, data: dict[str, Any]) -> FlowResult:
```

joostlek commented on homeassistant/components/husqvarna_automower/config_flow.py:
```diff
@@ -0,0 +1,57 @@
+        """Create an entry for the flow."""
+        return await self.async_step_finish(DOMAIN, data)
+
+    async def async_step_finish(self, unique_id, data) -> FlowResult:
```
You can merge this step with the step above as there is no real reason to split it out

joostlek commented on homeassistant/components/husqvarna_automower/const.py:
```diff
@@ -0,0 +1,20 @@
+NAME = "husqvarna_automower"
+DOMAIN = "husqvarna_automower"
+DOMAIN_DATA = f"{DOMAIN}_data"
+ISSUE_URL = "https://github.com/Thomas55555/husqvarna_automower"
```
```suggestion
```
We don't need this anymore as core will be the issue tracker to go :)

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+
+
+class AutomowerEntity(CoordinatorEntity[AutomowerDataUpdateCoordinator]):
+    """Defining the Automower Basic Entity."""
```
```suggestion
    """Defining the Automower base Entity."""
```

joostlek commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,100 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class AutomowerDataUpdateCoordinator(DataUpdateCoordinator):
```
```suggestion
class AutomowerDataUpdateCoordinator(DataUpdateCoordinator[None]):
```

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+    _attr_has_entity_name = True
+    _attr_should_poll = False
+
+    def __init__(self, coordinator, idx) -> None:
```
```suggestion
    def __init__(self, coordinator: AutomowerDataUpdateCoordinator, idx: int) -> None:
```
We love a well typed integration <3

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+    _attr_has_entity_name = True
+    _attr_should_poll = False
+
+    def __init__(self, coordinator, idx) -> None:
```
Why do we actually use a number to differentiate on the mowers? Could we maybe use some mower id instead?

Thomas55555 replied:
I thought it's the easiest way, becuase if we poll the API we get a list of mowers. So first mower is 0, second mower is 1 and so on.
Otherwise we would have to search in the dictionary of the mowers for the mower id

joostlek replied:
Can't we map it with mowerid: mower?

Thomas55555 replied:
As it is a dataclass now, it's easy to iterate through it. I think, it's okay now

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+
+        self._available = self.get_mower_attributes()["metadata"]["connected"]
+
+    def get_mower_attributes(self) -> dict:
```
```suggestion
    @property
    def mower_attributes(self) -> dict[str, Any]:
```

joostlek replied:
This way it's an property and can be accessed with `self.mower_attributes`

Thomas55555 replied:
I'm only a hobby programmer. I've always had the problem, that I had call the function. So otherwise the values stayed the same all the time. Is it really that simple solution?

joostlek replied:
you can see it as a semi function, you access it as property, but it calls the function but makes code way cleaner :)

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+        super().__init__(coordinator, context=idx)
+        self.idx = idx
+        self.mower = coordinator.session.data["data"][self.idx]
+        mower_attributes = self.get_mower_attributes()
```
```suggestion
```

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+        self.mower_name = mower_attributes["system"]["name"]
+        self.model_name = mower_attributes["system"]["model"]
+
+        self._available = self.get_mower_attributes()["metadata"]["connected"]
```
```suggestion
```
Looks like it isn't used

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+        self.mower = coordinator.session.data["data"][self.idx]
+        mower_attributes = self.get_mower_attributes()
+        self.mower_id = self.mower["id"]
+        self.mower_name = mower_attributes["system"]["name"]
+        self.model_name = mower_attributes["system"]["model"]
```
```suggestion
        self.mower_name = self.mower_attributes["system"]["name"]
        self.model_name = self.mower_attributes["system"]["model"]
```

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+        )
+
+    @property
+    def device_info(self) -> DeviceInfo:
```
You can initialize this in `_attr_device_info`

joostlek commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,72 @@
+        """Get the mower attributes of the current mower."""
+        return self.coordinator.session.data["data"][self.idx]["attributes"]
+
+    def datetime_object(self, timestamp) -> datetime.datetime | None:
```
I don't think this is a fitting place for this function. Would be better in a helpers.py or something along those lines. It could also be that there is a helper already in HA, but not that I know of.

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+"""Creates a vacuum entity for the mower."""
```
```suggestion
"""Husqvarna automower entity."""
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up vacuum platform."""
```
```suggestion
    """Set up lawn mower platform."""
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+        HusqvarnaAutomowerEntity(coordinator, idx)
+        for idx, ent in enumerate(coordinator.session.data["data"])
+    )
+    entity_platform.current_platform.get()
```
Why do we do this?

Thomas55555 replied:
actually not

joostlek replied:
what do you mean?

Thomas55555 replied:
That we don't need it and I have deleted it :)

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+class HusqvarnaAutomowerEntity(LawnMowerEntity, AutomowerEntity):
+    """Defining each mower Entity."""
+
+    _attr_name: str | None = None
```
```suggestion
    _attr_name = None
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+    def __init__(self, session, idx) -> None:
+        """Set up HusqvarnaAutomowerEntity."""
+        super().__init__(session, idx)
+        self._attr_unique_id = self.coordinator.session.data["data"][self.idx]["id"]
```
```suggestion
        self._attr_unique_id = session.data["data"][idx]["id"]
```

Thomas55555 replied:
It throws this error:
```
self._attr_unique_id = session.data["data"][idx]["id"]
                                       ~~~~~~~~~~^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

joostlek replied:
huh, it's the same session right?

Thomas55555 replied:
`self._attr_unique_id = session.session.data["data"][idx]["id"]` works.
I'm not sure, what's wrong there. But best thing would be to change it in:
`self._attr_unique_id = f"{self.mower_id}_lawn_mower"` so we can use in the future other unique_ids in the same way, like: `self._attr_unique_id = f"{self.mower_id}_battery"`

joostlek replied:
That looks even better!

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+    _attr_name: str | None = None
+    _attr_supported_features = SUPPORT_STATE_SERVICES
+
+    def __init__(self, session, idx) -> None:
```
```suggestion
    def __init__(self, session: AutomowerSession, idx: int) -> None:
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+    @property
+    def available(self) -> bool:
+        """Return True if the device is available."""
+        available = self.get_mower_attributes()["metadata"]["connected"]
+        return available
```
```suggestion
        return self.mower_attributes["metadata"]["connected"]
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+            "WAIT_UPDATING",
+            "WAIT_POWER_UP",
+        ]:
+            activity = LawnMowerActivity.PAUSED
```
```suggestion
            return LawnMowerActivity.PAUSED
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+        ]:
+            activity = LawnMowerActivity.PAUSED
+        if mower_attributes["mower"]["activity"] in ["MOWING", "LEAVING", "GOING_HOME"]:
+            activity = LawnMowerActivity.MOWING
```
```suggestion
            return LawnMowerActivity.MOWING
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+        """Return the state of the mower."""
+        mower_attributes = AutomowerEntity.get_mower_attributes(self)
+        if mower_attributes["mower"]["state"] in [
+            "PAUSED",
+            "WAIT_UPDATING",
+            "WAIT_POWER_UP",
```
Maybe worth making this a const list

MartinHjelmare replied:
The library she provide constants or enums for this. We shouldn't need to handle magic api values directly tas a user of the library.

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+        if mower_attributes["mower"]["activity"] in ["MOWING", "LEAVING", "GOING_HOME"]:
+            activity = LawnMowerActivity.MOWING
+        if (mower_attributes["mower"]["state"] == "RESTRICTED") or (
+            mower_attributes["mower"]["activity"] in ["PARKED_IN_CS", "CHARGING"]
```
Maybe worth making this a const list

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+        if (mower_attributes["mower"]["state"] == "RESTRICTED") or (
+            mower_attributes["mower"]["activity"] in ["PARKED_IN_CS", "CHARGING"]
+        ):
+            activity = LawnMowerActivity.DOCKED
```
```suggestion
            return LawnMowerActivity.DOCKED
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+    @property
+    def extra_state_attributes(self) -> dict:
+        """Return the specific state attributes of this mower."""
+        mower_attributes = AutomowerEntity.get_mower_attributes(self)
+        action = mower_attributes["planner"]["override"]["action"]
```
```suggestion
        action = self.mower_attributes["planner"]["override"]["action"]
```

joostlek commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+
+    async def async_start_mowing(self) -> None:
+        """Resume schedule."""
+        command_type = "actions"
+        payload = '{"data": {"type": "ResumeSchedule"}}'
+        try:
+            await self.coordinator.session.action(self.mower_id, payload, command_type)
+        except ClientResponseError as exception:
+            _LOGGER.error("Command couldn't be sent to the command que: %s", exception)
```
I think you can extract this functionality into a separate function where you only have to pass in the type, so for example
`execute_action("ResumeSchedule")`

MartinHjelmare replied:
The library client should expose easy to use methods. The library user shouldn't need to know to pass a `" ResumeSchedule"` string. Create different methods for different actions, resume_schedule`, `start_mowing` etc.

Thomas55555 replied:
I liked it, when the library was dumb, at could just pass any commands. This made it very flexible. But I changed the library. -> In my opinion it's now resolved. Let me know, if not.

joostlek commented on homeassistant/components/husqvarna_automower/strings.json:
```diff
@@ -0,0 +1,25 @@
+  "config": {
+    "step": {
+      "pick_implementation": {
+        "title": "[%key:common::config_flow::title::oauth2_pick_implementation%]"
```
This shouldn't be needed right?

MartinHjelmare replied:
Used by oauth2 helper.

joostlek commented on homeassistant/components/husqvarna_automower/strings.json:
```diff
@@ -0,0 +1,25 @@
+      "user_rejected_authorize": "[%key:common::config_flow::abort::oauth2_user_rejected_authorize%]"
+    },
+    "create_entry": {
+      "default": "[%key:common::config_flow::create_entry::authenticated%]"
```
This also shouldn't be needed

MartinHjelmare replied:
You can see what strings was added by the scaffold script here:
https://github.com/home-assistant/core/blob/b2b57c5f87a519af951a9d1ca8777f5ae3517b92/script/scaffold/generate.py#L175-L196

joostlek commented on homeassistant/components/husqvarna_automower/strings.json:
```diff
@@ -0,0 +1,25 @@
+      }
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_account%]",
+      "already_in_progress": "[%key:common::config_flow::abort::already_in_progress%]",
+      "oauth_error": "[%key:common::config_flow::abort::oauth2_error%]",
+      "missing_configuration": "[%key:common::config_flow::abort::oauth2_missing_configuration%]",
+      "authorize_url_timeout": "[%key:common::config_flow::abort::oauth2_authorize_url_timeout%]",
+      "no_url_available": "[%key:common::config_flow::abort::oauth2_no_url_available%]",
+      "user_rejected_authorize": "[%key:common::config_flow::abort::oauth2_user_rejected_authorize%]"
```
Do we actually throw all these aborts? I don't really think so

MartinHjelmare replied:
The oauth2 helper can use these errors so we need to keep it.

joostlek reviewed (changes requested):
Initial review, I checked at all the code, didn't check the tests yet. But I think this could give you a good start

MartinHjelmare commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+    async def async_dock(self) -> None:
+        """Parks the mower until next schedule."""
+        command_type = "actions"
+        payload = '{"data": {"type": "ParkUntilNextSchedule"}}'
```
This is too much protocol details. The library should abstract the protocol and make easy to use methods that the library user can use without knowing details about the Husqvarna service api.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+        payload = '{"data": {"type": "ParkUntilNextSchedule"}}'
+        try:
+            await self.coordinator.session.action(self.mower_id, payload, command_type)
+        except ClientResponseError as exception:
```
The library should preferably hide the transport details and raise library specific errors instead.

We shouldn't use the session directly as a library user. The library should provide a client that uses the transport session behind the scenes.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+    @property
+    def available(self) -> bool:
+        """Return True if the device is available."""
+        available = self.get_mower_attributes()["metadata"]["connected"]
```
I'd recommend creating dataclasses in the library from the parsed json data to make it easier for the library user to use the data returned from the api.

Thomas55555 commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,100 @@
+                is_fixable=False,
+                severity=IssueSeverity.WARNING,
+                translation_key="wrong_scope",
+            )
```
Yes, that's better. Should I remove this function and add reauth later?

joostlek replied:
Yes!

Thomas55555 commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+                "STOPPED",
+                "OFF",
+            ]
+        ) or mower_attributes["mower"]["activity"] in ERROR_ACTIVITIES:
```
yes

mikey0000 commented on tests/components/husqvarna_automower/test_lawn_mower.py:
```diff
@@ -0,0 +1,205 @@
+
+
+@pytest.mark.asyncio
+async def test_vacuum_state(hass: HomeAssistant) -> None:
```
```suggestion
async def test_lawn_mower_state(hass: HomeAssistant) -> None:
```

mikey0000 replied:
to add all the tests reference vacuum instead of lawn mower

Thomas55555 replied:
done

MartinHjelmare commented on homeassistant/components/husqvarna_automower/config_flow.py:
```diff
@@ -0,0 +1,35 @@
+
+    async def async_oauth_create_entry(self, data: dict[str, Any]) -> FlowResult:
+        """Create an entry for the flow."""
+        await self.async_set_unique_id(DOMAIN)
```
Why do we only allow one config entry?

Thomas55555 replied:
More are possible as well. I've added this function. The unique identifier is now the `auth_implemenation`, a combination of domain and Api-Key. As you can register for multiple Api-Keys in one account (with multiple mowers), there was the need to add the Api-Key to the unique_id of the lawn mower.
The disadavantage is now, if you register with a new Api-Key, there will be new entities now. Is this okay?

MartinHjelmare replied:
When you say api key, do you mean client secret? Doesn't the user account get access to all mowers of the user regardless of which application, ie which client id and secret, was set up to access the api?

MartinHjelmare replied:
Good. Is there a user id that we can query? If so we could use that as unique_id for the config entry.

Thomas55555 replied:
Yes we can :)
I change it.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,52 @@
+    return unload_ok
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
```
This function isn't used.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/__init__.py:
```diff
@@ -0,0 +1,52 @@
+    """Set up this integration using UI."""
+    hass.data.setdefault(DOMAIN, {})
+    implementation = await async_get_config_entry_implementation(hass, entry)
+    entry_dict = entry.as_dict()
```
We can get the token from the config entry data. We don't need to use `entry_as_dict`.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/const.py:
```diff
@@ -0,0 +1,33 @@
+OAUTH2_TOKEN = "https://api.authentication.husqvarnagroup.dev/v1/oauth2/token"
+
+# Platforms
+PLATFORMS: list[Platform] = [
```
`PLATFORMS` is only used in the init module. Please move it to that module. Only define the constants that are shared between multiple modules in the const module.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/coordinator.py:
```diff
@@ -0,0 +1,44 @@
+        self.session = aioautomower.AutomowerSession(
+            implementation.client_id, access_token
+        )
+        self.session.register_token_callback(
```
The config entry is updated by our helper when the token is refreshed. We should use that helper interface to refresh the token.

https://github.com/home-assistant/core/blob/b2b57c5f87a519af951a9d1ca8777f5ae3517b92/homeassistant/helpers/config_entry_oauth2_flow.py#L493-L503

Thomas55555 replied:
Right now, the helper function `async_ensure_token_valid()` is not implemented in the integration. Token updates are handled by the library. And the config entry is updated from the library if there is a new token.
What do you mean by "use the helper interface"?

MartinHjelmare replied:
The library client should use our helper session to authorize and refresh tokens.

See our library guide and use our scaffold script to get example code. Links here:

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#configuration-via-oauth2

See also eg the Google integration for a real implementation.

Thomas55555 replied:
Should be solved now.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/coordinator.py:
```diff
@@ -0,0 +1,44 @@
+        """Fetch data from Husqvarna."""
+        try:
+            await self.session.connect()
+        except Exception as error:
```
Please only catch expected specific exceptions.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,56 @@
+        """Call when entity about to be added to Home Assistant."""
+        await super().async_added_to_hass()
+        self.coordinator.session.register_data_callback(
+            lambda _: self.async_write_ha_state(), schedule_immediately=True
```
Please don't use a lambda. Define an entity method that calls `self.async_write_ha_state` and use the method as callback.

Thomas55555 replied:
Can you please help me with that in more detail?
I've tried this:
```
    async def async_added_to_hass(self) -> None:
        """Call when entity about to be added to Home Assistant."""
        await super().async_added_to_hass()
        self.coordinator.session.register_data_callback(
            self.async_method(), schedule_immediately=False
        )

    @callback
    async def async_method(self):
        self.async_write_ha_state()
```
But I'm getting this error:
`2023-09-13 21:42:38.853 ERROR (MainThread) [homeassistant] Error doing job: Exception in callback AutomowerEntity.async_method({'data': [{'attributes': {'battery': {'batteryPercent': 100}, 'calendar': {'tasks': [{...}]}, 'capabilities': {'headlights': True, 'position': True, 'stayOutZones': False, 'workAreas': False}, 'cuttingHeight': 1, ...}, 'id': 'b2bc3443-b31...-d6e408c53f1b', 'type': 'mower'}]})
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
TypeError: 'coroutine' object is not callable`

MartinHjelmare replied:
Pass the method when register the callback, don't call the callback. Ie remove the parenthesis.

A coroutine function is never a `@callback`. Does the update callback need to be a coroutine function or can it be a regular callback. Decide which one it is and adjust.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,56 @@
+
+    def __init__(self, coordinator: AutomowerDataUpdateCoordinator, idx: int) -> None:
+        """Initialize AutomowerEntity."""
+        super().__init__(coordinator, context=idx)
```
Where is the context needed?

Thomas55555 replied:
Don't know, where this is coming from. I've removed it.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,56 @@
+        """Initialize AutomowerEntity."""
+        super().__init__(coordinator, context=idx)
+        self.idx = idx
+        self.mower = coordinator.session.data["data"][self.idx]
```
` self.mower` is not used anywhere.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,56 @@
+        self.idx = idx
+        self.mower = coordinator.session.data["data"][self.idx]
+        self.mower_id = self.mower["id"]
+        self.mower_name = self.mower_attributes["system"]["name"]
```
`self.mower_name` isn't used anywhere outside this method. Just make it a local variable.

MartinHjelmare commented on homeassistant/components/husqvarna_automower/entity.py:
```diff
@@ -0,0 +1,56 @@
+        self.mower = coordinator.session.data["data"][self.idx]
+        self.mower_id = self.mower["id"]
+        self.mower_name = self.mower_attributes["system"]["name"]
+        self.model_name = self.mower_attributes["system"]["model"]
```
Same with `self.model_name`.

joostlek committed:
Bump pyenphase to v1.11.0 (#99941)

bdraco committed:
Avoid many hass.is_stopping calls in the discovery helper (#99929)

async_has_matching_flow is more likely to be True than hass.is_stopping

This does not make much difference but it was adding noise to a profile
that I am digging into to look for another issue

jbouwh committed:
Fix key error MQTT binary_sensor when no name is set (#99943)

Log entitty ID when instead of name

janiversen committed:
Bump pymodbus v.3.5.1 (#99940)

bdraco committed:
Switch a few ssdp calls to use get_lower (#99931)

get_lower avoids lower casing already lower-cased strings

bdraco committed:
Bump orjson to 3.9.7 (#99938)

emontnemery committed:
Bump hatasmota to 0.7.1 (#99818)

emontnemery committed:
Make WS command render_template not give up if initial render raises (#99808)

elmurato committed:
Add tests for Minecraft Server entry migration from v1 to v2 (#99954)

joostlek committed:
Use device class translations for Devolo Update entity (#99235)

cdce8p committed:
Update RestrictedPython to 6.2 (#99955)

bdraco committed:
Bump bleak to 0.21.1 (#99960)

bdraco committed:
Bump dbus-fast to 2.0.1 (#99894)

Kane610 committed:
Add UniFi device uptime and temperature sensors (#99307)

* Add UniFi device uptime and temperature sensors

* Add native_unit_of_measurement to temperature
Remove seconds and milliseconds from device uptime

Lash-L committed:
Bump python-roborock to 33.2 (#99962)

bump to 33.2

cdce8p committed:
Update black to 23.9.0 (#99965)

cdce8p committed:
Add black caching [ci] (#99967)

Co-authored-by: J. Nick Koston <nick@koston.org>

Roagert committed:
Add zdb5100 light to zwave_js (#97586)

* added zdb5100 light

* added light to zdb5100

* Update tests/components/zwave_js/conftest.py

agree

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Update tests/components/zwave_js/conftest.py

agree

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Rename logic_group_zdb5100_light_state.json to logic_group_zdb5100_state.json

name change

* Update tests/components/zwave_js/test_light.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Update test_light.py

updated test and state

* Update test_light.py

incorrect endpoint

* changed the state

---------

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

gjohansson-ST committed:
Bump pytrafikverket to 0.3.6 (#99869)

* Bump pytrafikverket to 0.3.6

* Fix config flow names

* str

bdraco committed:
Change SSDP discovery scan interval to 10 minutes (#99975)

* Change SSDP discovery scan interval to 10 minutes

The first version used a scan interval of 1 minute which we
increased to 2 minutes because it generated too much traffic.

We kept it at 2 minutes because Sonos historicly needed to
get SSDP discovery to stay alive. This is no longer the case
as Sonos has multiple ways to keep from going unavailable:

- mDNS support was added
- We now listen for SSDP alive and good bye all the time
- Each incoming packet from the device keeps it alive now
- We probe when we think the device might be offline

This means it should no longer be necessary to have such a
frequent scan which is a drag on all devices on the network
since its multicast

* adjust tests

janiversen committed:
Bump pymodbus to v3.5.2 (#99988)

joostlek committed:
Add config flow to WAQI (#98220)

* Migrate WAQI to aiowaqi library

* Migrate WAQI to aiowaqi library

* Migrate WAQI to aiowaqi library

* Add config flow to WAQI

* Finish config flow

* Add tests

* Add tests

* Fix ruff

* Add issues on failing to import

* Add issues on failing to import

* Add issues on failing to import

* Add importing issue

* Finish coverage

* Remove url from translation string

* Fix feedback

* Fix feedback

Lash-L committed:
Add binary sensors to Roborock (#99990)

* init binary sensors commit

* add binary sensors

* fix test

* Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

janiversen committed:
Deprecate CLOSE_COMM_ON_ERROR (#99946)

joostlek committed:
Add entity translations to withings (#99194)

* Add entity translations to Withings

* Add entity translations to Withings

joostlek committed:
Decouple Withings sensor tests from yaml (#99691)

* Decouple Withings sensor tests from yaml

* Fix feedback

* Add pytest fixture

* Update tests/components/withings/test_sensor.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

* Update snapshots

* Update snapshots

---------

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

eifinger committed:
Bump pywaze to 0.4.0 (#99995)

bump pywaze from 0.3.0 to 0.4.0

bdraco committed:
Bump bluetooth-auto-recovery to 1.2.3 (#99979)

fixes #99977

elmurato committed:
Fix late review findings in Minecraft Server (#99865)

joostlek committed:
Use shorthand attributes in Vicare (#99915)

joostlek committed:
Restore airtouch4 codeowner (#99984)

bouwew committed:
Bump plugwise to v0.32.2 (#99973)

* Bump plugwise to v0.32.2

* Adapt number.py to the backend updates

* Update related test-cases

* Update plugwise test-fixtures

* Update test_diagnostics.py

bdraco committed:
Bump yalexs-ble to 2.3.0 (#100007)

bdraco committed:
Switch ESPHome Bluetooth to use loop.create_future() (#100010)

bdraco committed:
Switch config_entries to use loop.create_future() (#100011)

joostlek committed:
Use snapshot assertion in Plugwise diagnostic test (#100008)

* Use snapshot assertion in Plugwise diagnostic test

* Use snapshot assertion in Plugwise diagnostic test

joostlek committed:
Move static attributes outside of ws66i constructor (#99922)

Move static attributes outside of ws66i cosntructor

joostlek committed:
Fix renamed code owner for Versasense (#99976)

Fix renamed code owner

dieselrabbit committed:
Bump screenlogicpy to v0.9.0 (#92475)

Co-authored-by: J. Nick Koston <nick@koston.org>

Lash-L committed:
Add new sensors to Roborock (#99983)

* Add 3 new sensor types

* add state options for dock error

* add unit of measurement

bdraco committed:
Bump zeroconf to 0.103.0 (#100012)

bdraco committed:
Bump aiodiscover to 1.5.1 (#100020)

changelog: https://github.com/bdraco/aiodiscover/compare/v1.4.16...v1.5.1

elmurato committed:
Add sensor entity descriptions in Minecraft Server (#99971)

* Add sensor entity descriptions

* Fix review findings

* Fix type of value function to avoid inline lambda if conditions and add attribute function to avoid extra sensor entity class

* Correct name of binary sensor base entity

* Simplify adding of entities in platforms

* Do not use keyword arguments while adding entities

bouwew committed:
Add Plugwise temperature_offset number (#100029)

Add temperature_offset number

bouwew committed:
Enable strict typing in Plugwise (#100033)

Add plugwise to .strict-typing

fender4645 committed:
Add debug message to doods (#100002)

* Debug message if no detections found or
no output file configured

* fix formatting

* black

---------

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

bdraco committed:
Speed up serializing event messages (#100017)

bdraco committed:
Avoid json_decoder_fallback in /api/states (#100018)

dgomes committed:
Always update unit of measurement of the utility_meter on state change (#99102)

cdce8p committed:
Update bthome-ble to 3.1.1 (#100042)

mkmer committed:
Add diagnostic platform to Honeywell (#100046)

Add diagnostic platform

mkmer committed:
Raise HomeAssistantError/ValueError for service calls in Honeywell (#100041)

vfreex committed:
Make homekit RTP/RTCP source ports more deterministic (#99989)

joostlek committed:
Improve Withings config flow tests (#99697)

* Decouple Withings sensor tests from yaml

* Improve Withings config flow tests

* Improve Withings config flow tests

* Fix feedback

* Rename CONF_PROFILE to PROFILE

timmo001 committed:
Bump systembridgeconnector to 3.8.2 (#100051)

Update systembridgeconnector to 3.8.2

joostlek committed:
Add device to OpenUV (#100027)

jbouwh committed:
Uer hass.loop.create_future() for MQTT client (#100053)

jbouwh committed:
Use hass.loop.create_future() for bluetooth (#100054)

timmo001 committed:
Add System Bridge notifications (#82318)

* System bridge notifications

Add notify platform

Add file to coverage

Restore and fix lint after rebase

Cleanup

Use entity to register notify service

Fix pylint

Update package to 3.6.0 and add audio actions

Update package to fix conflict

Remove addition

* Run pre-commit run --all-files

* Update homeassistant/components/system_bridge/notify.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Format

* Fix

* Remove duplicate import

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

bdraco committed:
Create recorder futures with loop.create_future() (#100049)

jbouwh committed:
Use hass.loop.create_future() in envisalink (#100057)

jpbede committed:
Remove unnecessary argument from discovergy coordinator (#100058)

ms264556 committed:
Address ruckus_unleashed late review (#99411)

bdraco committed:
Bump aiohomekit to 3.0.3 (#100047)

chemelli74 committed:
Bump aiovodafone to 0.2.0 (#100062)

bump aiovodafone to 0.2.0

Noltari committed:
Add water heater support to Airzone (#98401)

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Add __slots__ to translation cache (#100069)

bdraco committed:
Bump zeroconf to 0.104.0 (#100068)

bdraco committed:
Bump flux_led to 1.0.4 (#100050)

dependabot[bot] committed:
Bump tibdex/github-app-token from 1.8.0 to 1.8.2 (#99434)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bdraco committed:
Bump dbus-fast to 2.2.0 (#100076)

matrixd2 committed:
Add yolink siren battery entity (#99310)

gjohansson-ST committed:
Add sensors to Trafikverket Camera (#100078)

* Add sensors to Trafikverket Camera

* Remove active

* Fix test len

gjohansson-ST committed:
Fix missed name to translation key in Sensibo (#100080)

gjohansson-ST committed:
Add binary sensors to Trafikverket Camera (#100082)

gjohansson-ST committed:
Fix incorrect docstring in TV Camera sensor test (#100083)

jimmyd-be committed:
Renson number entity (#99358)

* Starting number sensor

* Filter change config

* Add translation to number entity

* add number entity to .coveragerc

* Moved has_entity_name to description + changed name of entity

* Add self.coordinator.async_request_refresh() after changing value

* Add device calss and unit of measurement to number entity

mikey0000 committed:
Add unit tests for sensors Electric Kiwi (#97723)

* add unit tests for sensors

* newline long strings

* unit test check and move time

* rename entry to entity

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* add types to test

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* fix newlined f strings

* remove if statement

* add some more explaination

* Update datetime

Co-authored-by: Robert Resch <robert@resch.dev>

* Simpler time update

Co-authored-by: Robert Resch <robert@resch.dev>

* add missing datetime import

* Update docustring - grammar

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* address comments and issues raised

* address docstrings too long

* Fix docstring

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
Co-authored-by: Robert Resch <robert@resch.dev>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

chemelli74 committed:
Remove Comelit alarm data retrieval (#100067)

fix: remove alarm data retrieval

greiginsydney committed:
Remove duplicated word in enphase description text (#100098)

joostlek committed:
Use shorthand attributes in ZAMG (#99925)

Co-authored-by: Robert Resch <robert@resch.dev>

engrbm87 committed:
Move smtp constants to const.py (#99542)

jbouwh committed:
Add TYPE_CHECKING condition on type assertions for mqtt (#100107)

Add TYPE_CHECKING condition on type assertions

jpbede committed:
Also handle DiscovergyClientError as UpdateFailed (#100038)

* Also handle DiscovergyClientError as UpdateFailed

* Change AccessTokenExpired to InvalidLogin

* Also add DiscovergyClientError to config flow and tests

ratsept committed:
Fix Soma cover tilt (#99717)

cdce8p committed:
Update black to 23.9.1 (#100108)

joostlek committed:
Use json to store Withings test data fixtures (#99998)

* Decouple Withings sensor tests from yaml

* Improve Withings config flow tests

* Improve Withings config flow tests

* Fix feedback

* Use fixtures to store Withings testdata structures

* Use fixtures to store Withings testdata structures

* Use JSON

* Fix

* Use load_json_object_fixture

joostlek committed:
Decouple Withings binary sensor test from YAML (#100120)

emontnemery committed:
Fix TriggerEntity.async_added_to_hass (#100119)

puddly committed:
Cache device trigger info during ZHA startup (#99764)

* Do not connect to the radio hardware within `_connect_zigpy_app`

* Make `connect_zigpy_app` public

* Create radio manager instances from config entries

* Cache device triggers on startup

* reorg zha init

* don't reuse gateway

* don't nuke yaml configuration

* review comments

* Fix existing unit tests

* Ensure `app.shutdown` is called, not just `app.disconnect`

* Revert creating group entities and device registry entries early

* Add unit tests

---------

Co-authored-by: David F. Mulcahey <david.mulcahey@icloud.com>

bdraco committed:
Bump pyunifiprotect to 4.20.0 (#100092)

bdraco committed:
Bump zeroconf to 0.105.0 (#100084)

bdraco committed:
Cleanup isinstance checks in zeroconf (#100090)

Kane610 committed:
UniFi streamline loading platforms (#100071)

* Streamline loading platforms

* Move platform registration logic to UnifiController class

Noltari committed:
Remove AEMET daily precipitation sensor test (#100118)

emontnemery committed:
Fix ScrapeSensor.async_added_to_hass (#100125)

starkillerOG committed:
Fix inverse naming of function in Reolink (#100113)

emontnemery committed:
Bump hatasmota to 0.7.2 (#100129)

bdraco committed:
Guard expensive debug logging with isEnabledFor in alexa (#100137)

tjhorner committed:
Add Apple WeatherKit integration (#99895)

bdraco committed:
Migrate mobile_app to use json helper (#100136)

bdraco committed:
Bump zeroconf to 0.107.0 (#100134)

changelog: https://github.com/python-zeroconf/python-zeroconf/compare/0.105.0...0.107.0

niro1987 committed:
Remove code owner Verisure (#100145)

joostlek committed:
Decouple Withings webhook tests from YAML (#100143)

puddly committed:
Migrate internal ZHA data to a dataclasses (#100127)

* Cache device triggers on startup

* reorg zha init

* don't reuse gateway

* don't nuke yaml configuration

* review comments

* Add unit tests

* Do not cache device and entity registries

* [WIP] Wrap ZHA data in a dataclass

* [WIP] Get unit tests passing

* Use a helper function for getting the gateway object to fix annotations

* Remove `bridge_id`

* Fix typing issues with entity references in group websocket info

* Use `Platform` instead of `str` for entity platform matching

* Use `get_zha_gateway` in a few more places

* Fix flaky unit test

* Use `slots` for ZHA data

Co-authored-by: J. Nick Koston <nick@koston.org>

---------

Co-authored-by: David F. Mulcahey <david.mulcahey@icloud.com>
Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Bump zeroconf to 0.108.0 (#100148)

Kane610 committed:
Fix devices not always reporting IP - bump aiounifi to v62 (#100149)

jpbede committed:
Refactor discovergy config flow test to use parametrize (#100115)

* Refactor discovergy config flow test to use parametrize

* Formatting

* Implement code review sugesstions

joostlek committed:
Split Withings common file out to their own file (#100150)

* Split common out in logical pieces

* Split common out in logical pieces

* Split common out in logical pieces

Noltari committed:
Use library for condition/wind direction conversions (#100117)

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

jpbede committed:
Let the discovergy config flow test end with create entry (#100153)

bramkragten committed:
Update frontend to 20230911.0 (#100139)

Thomas55555 commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,122 @@
+        action = self.mower_attributes["planner"]["override"]["action"]
+        action = action.lower() if action is not None else action
+        return {
+            "action": action,
```
Possible values are `NOT_ACTIVE`, `FORCE_PARK`, `FORCE_MOW`

MartinHjelmare replied:
What does `NOT_ACTIVE` mean?

MartinHjelmare replied:
Yes, I think we could add it as a sensor with device class enum and call it eg `schedule_override` or similar.

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

raman325 committed:
Handle disconnects in zwave_js repair flow (#99964)

* Handle disconnects in zwave_js repair flow

* Combine logic to reduce LoC

* only check once

puddly committed:
Bump ZHA dependencies (#100156)

mib1185 committed:
Fix AVM Fritz!Tools update entity (#100151)

* move update entity to coordinator

* fix tests

bdraco committed:
Avoid probing ipp printers for unique_id when it is available via mdns (#99982)

* Avoid probing ipp printers for unique_id when it is available via mdns

We would always probe the device in the ipp flow and than
abort if it was already configured. We avoid the probe for
most printers.

* dry

* coverage

* fix test

* add test for updating host

bdraco committed:
Bump dbus-fast to 2.4.0 (#100158)

jpbede committed:
Move airly coordinator to its own file (#99545)

raman325 committed:
Bump zwave-js-server-python to 0.51.2 (#100159)

allenporter committed:
Add configuration flow to Todoist integration (#100094)

* Add config flow to todoist

* Fix service calls for todoist

* Fix configuration entry test setup

* Bump test coverage to 100%

* Apply pr feedback

bdraco committed:
Bump dbus-fast to 2.6.0 (#100163)

changelog: https://github.com/Bluetooth-Devices/dbus-fast/compare/v2.4.0...v2.6.0

emontnemery committed:
Adjust tasmota sensor device class and icon mapping (#100168)

Noltari committed:
Add missing AEMET wind gust speed (#100157)

alexyao2015 committed:
Add Life360 Location Update Button (#99559)

Co-authored-by: Robert Resch <robert@resch.dev>
Co-authored-by: alexyao2015 <alexyao2015@users.noreply.github.com>

joostlek committed:
Use shorthand attributes in Smart meter texas (#99838)

Co-authored-by: Robert Resch <robert@resch.dev>

emontnemery committed:
Bump hatasmota to 0.7.3 (#100169)

dependabot[bot] committed:
Bump tibdex/github-app-token from 1.8.2 to 2.0.0 (#100099)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

vpathuis committed:
Bump Ultraheat to version 0.5.7 (#100172)

janiversen committed:
Bring modbus naming in sync with standard (#99285)

emontnemery committed:
Remove unnecessary block use of pylint disable in components a-o (#100190)

janiversen committed:
Remove default from deprecated close_comm_on_error (#100188)

joostlek committed:
Use shorthand attributes in Wiffi (#99919)

joostlek committed:
Use shorthand attributes in Songpal (#99849)

joostlek committed:
Use shorthand attributes in Squeezebox (#99863)

joostlek committed:
Use shorthand attributes in Twinkly (#99891)

joostlek committed:
Use shorthand attributes in Smarttub (#99839)

jbouwh committed:
Fix entity name attribute on mqtt entity is not removed on update (#100187)

Fix entity name attribute is not remove on update

joostlek committed:
Use shorthand attributes in Vera (#99893)

emontnemery committed:
Simplify WS command entity/source (#99439)

ludeeus committed:
Bump hass-nabucasa from 0.70.0 to 0.71.0 (#100193)

Bump hass-nabucasa from 0.70.0 to 0.71.1

mdegat01 committed:
Fix addon slug validation (#100070)

* Fix addon slug validation

* Don't redefine compile

LaStrada committed:
Airthings BLE unique id migration (#99832)

* Fix sensor unique id

* Add sensor identifiers

* Migrate entities to new unique id

* Fix linting issues

* Fix crash when migrating entity fails

* Change how entities are migrated

* Remve debug logging

* Remove unneeded async

* Remove migration code from init file

* Add migration code to sensor.py

* Adjust for loops to improve speed

* Bugfixes, improve documentation

* Remove old comment

* Remove unused function parameter

* Address PR feedback

* Add tests

* Improve tests and test data

* Refactor test

* Update logger level

Co-authored-by: J. Nick Koston <nick@koston.org>

* Adjust PR comments

* Address more PR comments

* Address PR comments and adjust tests

* Fix PR comment

---------

Co-authored-by: J. Nick Koston <nick@koston.org>

janiversen committed:
Make modbus retry fast on read errors (#99576)

* Fast retry on read errors.

* Review comments.

janiversen committed:
Remove modbus pragma no cover and solve nan (#99221)

* Remove pragma no cover.

* Ruff !

* Review comments.

* update test.

* Review.

* review.

* Add slave test.

emontnemery committed:
Remove unnecessary block use of pylint disable in onvif (#100194)

joostlek committed:
Use shorthand attribute in Bloomsky (#100203)

bdraco committed:
Use shorthand attributes in geniushub sensor (#100208)

bdraco committed:
Use shorthand attributes in buienradar camera (#100205)

bdraco committed:
Use shorthand attributes in garadget cover (#100207)

emontnemery committed:
Enable asyncio debug mode in tests (#100197)

joostlek committed:
Use shorthand attributes in Isy994 (#100209)

joostlek committed:
Use shorthand attributes in Point (#100214)

emontnemery committed:
Improve type annotations in websocket_api tests (#100198)

joostlek committed:
Use shorthand attributes in Universal (#100219)

joostlek committed:
Use shorthand attributes in Yamaha Musiccast (#100220)

joostlek committed:
Use shorthand attributes in SPC (#100217)

rklomp committed:
Add Entity Descriptions to SMA integration (#58707)

Co-authored-by: J. Nick Koston <nick@koston.org>

jbouwh committed:
Add intial property to imap_content event data (#100171)

* Add initial property to imap event data

* Simplify loop

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* MyPy

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

joostlek committed:
Use shorthand attributes in Smartthings (#100215)

emontnemery committed:
Remove unnecessary pylint disable in tado (#100196)

hahn-th committed:
Bump homematicip_cloud to 1.0.15 (#99387)

bdraco committed:
Set TriggerBaseEntity device_class in init (#100216)

uvjustin committed:
Update Stream logging on EVENT_LOGGING_CHANGED (#99256)

jrieger committed:
Use more common translations (#100135)

bdraco committed:
Use more shorthand attributes in huawei_lte binary_sensor (#100211)

bdraco committed:
Use more shorthand properties in homematicip_cloud (#100210)

bdraco committed:
Set roku media player device class in constructor (#100225)

joostlek committed:
Use shorthand attributes in NWS (#99620)

SirGoodenough committed:
Update Deprecated Selector Syntax (#99308)

dependabot[bot] committed:
Bump sigstore/cosign-installer from 3.1.1 to 3.1.2 (#99563)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bdraco committed:
Use short hand entity_registry_enabled_default in nws (#100227)

* Use short hand entity_registry_enabled_default in nws

see https://github.com/home-assistant/core/pull/95315

* Update homeassistant/components/nws/sensor.py

joostlek committed:
Add entity translations to WLED (#99056)

mkmer committed:
Clean up device registry for climate devices that no longer exist in Honeywell (#100072)

bdraco committed:
Use short handle attributes for device class in netatmo cover (#100228)

emontnemery committed:
Improve typing of entity.entity_sources (#99407)

* Improve typing of entity.entity_sources

* Calculate entity info source when generating WS response

* Adjust typing

* Update tests

jbouwh committed:
Use shorthanded attributes for MQTT cover (#100230)

emontnemery committed:
Remove unnecessary block use of pylint disable in components p-z (#100192)

joostlek committed:
Introduce base class for Trafikverket camera (#100114)

* Introduce base class for Trafikverket camera

* fix feedback

* Fix feedback

bdraco committed:
Set dynalite cover device class in constructor (#100232)

joostlek committed:
Clean up variables in Soundtouch (#99859)

emontnemery committed:
Improve Entity._suggest_report_issue (#100204)

allenporter committed:
Remove the uniqueid from todoist (#100206)

Noltari committed:
Add meteoclimatic sensor statistics (#100186)

teharris1 committed:
Support for Insteon 4 button KeypadLink device (#100132)

joostlek committed:
Remove restore functionality in Speedtest.net (#96950)

jbouwh committed:
User shorthand attr for mqtt alarm_control_panel (#100234)

jbouwh committed:
Use shorthand attr for mqtt assumed_state (#100241)

jrieger committed:
Add strict typing to GPSD (#100030)

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

scop committed:
Upgrade ruff to 0.0.289 (#100238)

bdraco committed:
Update ecobee zeroconf/homekit discovery (#100091)

joostlek committed:
Use shorthand attributes in VLC telnet (#99916)

* Use shorthand attributes in VLC telnet

* Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* fix mypy

* Attempt 3

---------

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Bump async-upnp-client to 0.35.1 (#100248)

starkillerOG committed:
Bump pynetgear to 0.10.10 (#100242)

bdraco committed:
Disable always responding to all SSDP M-SEARCH requests with the root device (#100224)

cgarwood committed:
Bump pyenphase to 1.11.2 (#100249)

* Bump pyenphase to 1.11.1

* Apply suggestions from code review

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

Lash-L committed:
Fix incorrect off peak translation key for Roborock (#100246)

fix incorrect translation key

chemelli74 committed:
SamsungTV: Add unique_id for when missing (legacy models) (#96829)

* Add unique_id for when missing (legacy models)

* add comment

* update tests, thx @epenet

cdce8p committed:
Update jsonpath to 0.82.2 (#100252)

cdce8p committed:
Fix pylint config warning (#100251)

cgarwood committed:
Bump pyenphase to 1.11.3 (#100255)

tjhorner committed:
Update apple_weatherkit to 1.0.2 (#100254)

Lash-L committed:
Bump python-roborock to 0.34.0 (#100236)

jhollowe committed:
Correct Venstar firmware version to use device's FW version instead of API version (#98493)

tjhorner committed:
Address weatherkit late review comments (#100265)

* Address review comments from original weatherkit PR

* Use .get() for optional fields

dependabot[bot] committed:
Bump docker/login-action from 2.2.0 to 3.0.0 (#100264)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

akx committed:
Bump sensirion-ble to 0.1.1 (#100271)

Bump to sensirion-ble==0.1.1

Fixes akx/sensirion-ble#6

Refs https://github.com/home-assistant/core/issues/93678#issuecomment-1694522112

emontnemery committed:
Tweak entity/source WS command handler (#100272)

joostlek committed:
Test speedtest.net config entry lifecycle (#100280)

emontnemery committed:
Future proof assist_pipeline.Pipeline (#100277)

emontnemery committed:
Improve type hint in entity_registry (#100278)

Olen committed:
Add diagnostics to Twinkly (#100146)

jbouwh committed:
Use shorthanded attrs for yamaha_musiccast select (#100273)

jpbede committed:
Move sms coordinators to their own file (#100276)

eifinger committed:
Limit waze_travel_time to 1 call every 0.5s (#100191)

jbouwh committed:
Use hass.loop.create_future() in zha (#100056)

* Use hass.loop.create_future() in zha

* Remove not needed method

jbouwh committed:
Use shorthanded attrs for vera sensor (#100269)

Co-authored-by: Franck Nijhof <git@frenck.dev>

joostlek committed:
Remove NZBGet configurable scan interval (#98869)

frenck committed:
Update syrupy to 4.5.0 (#100283)

frenck committed:
Update mutagen to 1.47.0 (#100284)

jbouwh committed:
Use shorthand attrs for xbox base_sensor (#100290)

starkillerOG committed:
Netgear catch no info error (#100212)

joostlek committed:
Remove profile from Withings config flow (#100202)

* Remove profile from Withings config flow

* Add config flow migration

* Add config flow migration

* Remove datamanager profile

* Remove datamanager profile

* Add manufacturer

* Remove migration

* Remove migration

* Fix feedback

cgarwood committed:
Bump pyenphase to 1.11.4 (#100288)

jbouwh committed:
Use shorthand attrs in iaqualink (#100281)

* Use shorthand attrs in iaqualink

* Use super

* Update homeassistant/components/iaqualink/light.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Remove self

* More follow ups

* Remove cast and type check

* Update homeassistant/components/iaqualink/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jbouwh committed:
Use shorthand attrs for velux (#100294)

* Use shorthand attrs for velux

* Update homeassistant/components/velux/cover.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* black

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

bdraco committed:
Fix missing super async_added_to_hass in lookin (#100296)

emontnemery committed:
Replace StateMachine._domain_index with a UserDict (#100270)

* Replace StateMachine._domain_index with a UserDict

* Access the UserDict's backing dict directly

* Optimize

joostlek committed:
Add options flow to Withings (#100300)

jrieger committed:
Allow setting the elevation in `set_location` (#99978)

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

emontnemery committed:
Improve UserDict in device and entity registries (#100307)

bdraco committed:
Bump yalexs to 1.9.0 (#100305)

frenck committed:
Update debugpy to 1.8.0 (#100311)

jbouwh committed:
Use shorthand attrs in template integration (#100301)

balloob committed:
Update Roborock entity categories (#100316)

Kane610 committed:
Remove legacy UniFi PoE client clean up (#100318)

bdraco committed:
Use short hand attributes in onvif camera (#100319)

see #95315

Quentame committed:
Add some tests to Freebox (#99755)

Danielhiversen committed:
Update Mill library to 0.11.5, handle rate limiting (#100315)

bdraco committed:
Bump dbus-fast to 2.7.0 (#100321)

bdraco committed:
Use cached_property in entity_registry (#100302)

jpbede committed:
Use snapshot assertion for nexia diagnostics test (#100328)

bdraco committed:
Bump python-amcrest to 1.9.8 (#100324)

jpbede committed:
Add missing sms coordinator to .coveragerc (#100327)

bdraco committed:
Use cached_property in device registry (#100309)

frenck committed:
Update sentry-sdk to 1.31.0 (#100293)

jbouwh committed:
Use shorthand atts for met_eireann (#100335)

jbouwh committed:
Shorthanded attrs for met integration (#100334)

Thomas55555 committed:
remove extra state attributes

Thomas55555 committed:
remove lambda

Thomas55555 committed:
remove aioautomower folder

Thomas55555 commented:
I messed this PR up. Is it okay, when I close it and open a new one?

MartinHjelmare commented:
Yes, please close and start a fresh branch and PR. Thanks!

Thomas55555 commented on homeassistant/components/husqvarna_automower/lawn_mower.py:
```diff
@@ -0,0 +1,124 @@
+            ]
+        ) or mower_attributes["mower"]["activity"] in ERROR_ACTIVITIES:
+            activity = LawnMowerActivity.ERROR
+        return LawnMowerActivity(activity)
```
I've gone back to the state like it was, because in my opinion it's not very likely, that it changes.
